### ISSUE TITLE
[BUG] A potential DHCP renewal issue discovered in kube-vip for DHCP-based load balancers

### ISSUE THREAD

**Issue by starbops** [harvester/harvester#6563]
**Describe the bug**

TL;DR: The DHCP renewal operation always fails when using the DHCP-IPAM VM load balancer. Though the subsequent rebinding operation has succeeded, and the load balancer retains its IP address, it's still worth our attention.

---

When a client obtains an IP address from a DHCP server, the server specifies a lease period, which includes two critical time points:

- **T1 (renewal time)**: At 50% of the lease duration, the client enters the renewal phase and sends a DHCP REQUEST to the original DHCP server to renew the lease.
- **T2 (rebinding time)**: At 87.5% of the lease duration, the client enters the rebind phase if the original DHCP server doesn't respond.

To renew the lease with the original DHCP server, the client needs to create a listening UDP socket on port 68, which is already occupied by the other process. That causes the renewal procedure to fail at an early stage.

Assume we have a running virtual machine:

```shell
$ kubectl get vmis
NAME      AGE     PHASE     IP             NODENAME                 READY
test-vm   5h21m   Running   172.19.32.56   harvester-vm-1-default   True
```

And a DHCP IPAM load balancer was created for SSH accessing the virtual machine:

```shell
$ kubectl get lbs
NAME          DESCRIPTION   WORKLOADTYPE   IPAM   ADDRESS        AGE
test-vm-ssh                 vm             dhcp   172.19.31.96   4h58m
```

Kube-vip created a `macvlan` interface for the DHCP client to negotiate with the DHCP server. The retrieved IP address was then assigned to the `mgmt-br` interface. Logs generated from the kube-vip pod during the DHCP IPAM load balancer creation:

```
# Logs from harvester-system/kube-vip-rjllh pod
time="2024-09-13T03:53:35Z" level=info msg="Creating new macvlan interface for DHCP [vip-0e80e041]"
time="2024-09-13T03:53:35Z" level=info msg="Generated mac: 00:00:6C:09:4c:d0"
time="2024-09-13T03:53:35Z" level=info msg="New interface [vip-0e80e041] mac is 00:00:6c:09:4c:d0"
time="2024-09-13T03:53:39Z" level=info msg="(svcs) adding VIP [172.19.31.96] via mgmt-br for [default/test-vm-ssh]"
time="2024-09-13T03:53:39Z" level=info msg="[service] synchronised in 3210ms"
time="2024-09-13T03:53:39Z" level=warning msg="(svcs) already found existing address [172.19.31.96] on adapter [mgmt-br]"
time="2024-09-13T03:53:42Z" level=warning msg="Re-applying the VIP configuration [172.19.31.96] to the interface [mgmt-br]"
```

The created `macvlan` interface on the node:

```shell
harvester-vm-0-default:~ # ip link show vip-0e80e041
98: vip-0e80e041@mgmt-br: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP mode DEFAULT group default
    link/ether 00:00:6c:09:4c:d0 brd ff:ff:ff:ff:ff:ff
```

The load balancer's IP address was bound to the `mgmt-br` interface, which can be found on the node below:

```shell
harvester-vm-0-default:~ # ip addr show mgmt-br
5: mgmt-br: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
    link/ether 52:54:00:01:00:01 brd ff:ff:ff:ff:ff:ff
    inet 172.19.31.224/24 brd 172.19.31.255 scope global mgmt-br
       valid_lft forever preferred_lft forever
    inet 172.19.31.240/32 scope global mgmt-br
       valid_lft forever preferred_lft forever
    inet 172.19.31.96/32 scope global mgmt-br
       valid_lft forever preferred_lft forever
```

Wait for a while (til the lease is about to expire), the failed renewal and successful rebinding can be observed from the kube-vip pod's log:

```
# Logs from harvester-system/kube-vip-rjllh pod
time="2024-09-13T07:01:09Z" level=error msg="renew failed, error: create renew client failed, error: unable to apply option: unable to start listening UDP port: listen udp4 172.19.31.96:68: bind: address already in use, server ip: 172.19.31.1"
time="2024-09-13T07:23:39Z" level=info msg="rebind, lease: &{Offer:DHCPv4(xid=0xd8699f2d hwaddr=00:00:6c:09:4c:d0 msg_type=OFFER, your_ip=172.19.31.96, server_ip=172.19.31.1) ACK:DHCPv4(xid=0xd8699f2d hwaddr=00:00:6c:09:4c:d0 msg_type=ACK, your_ip=172.19.31.96, server_ip=172.19.31.1) CreationTime:2024-09-13 07:23:39.249844537 +0000 UTC m=+18458.023101337}"
time="2024-09-13T07:53:39Z" level=error msg="renew failed, error: create renew client failed, error: unable to apply option: unable to start listening UDP port: listen udp4 172.19.31.96:68: bind: address already in use, server ip: 172.19.31.1"
time="2024-09-13T08:16:09Z" level=info msg="rebind, lease: &{Offer:DHCPv4(xid=0x628fd85b hwaddr=00:00:6c:09:4c:d0 msg_type=OFFER, your_ip=172.19.31.96, server_ip=172.19.31.1) ACK:DHCPv4(xid=0x628fd85b hwaddr=00:00:6c:09:4c:d0 msg_type=ACK, your_ip=172.19.31.96, server_ip=172.19.31.1) CreationTime:2024-09-13 08:16:09.291550658 +0000 UTC m=+21608.064807428}"
```

No packets were emitted since the renewal procedure failed at the very beginning point. Below are the packets observed during the DHCP rebinding process on the node where the VIP address was bound:

```
harvester-vm-0-default:~ # tcpdump -ennvi mgmt-bo port 67
tcpdump: listening on mgmt-bo, link-type EN10MB (Ethernet), snapshot length 262144 bytes
08:16:09.275262 00:00:6c:09:4c:d0 > ff:ff:ff:ff:ff:ff, ethertype IPv4 (0x0800), length 342: (tos 0x0, ttl 64, id 0, offset 0, flags [none], proto UDP (17), length 328)
    0.0.0.0.68 > 255.255.255.255.67: BOOTP/DHCP, Request from 00:00:6c:09:4c:d0, length 300, xid 0x628fd85b, Flags [none]
          Client-Ethernet-Address 00:00:6c:09:4c:d0
          Vendor-rfc1048 Extensions
            Magic Cookie 0x63825363
            Requested-IP (50), length 4: 172.19.31.96
            DHCP-Message (53), length 1: Discover
            Parameter-Request (55), length 4:
              Subnet-Mask (1), Default-Gateway (3), Domain-Name (15), Domain-Name-Server (6)
            MSZ (57), length 2: 1500
08:16:09.276215 4a:1f:7b:fb:b6:8e > 00:00:6c:09:4c:d0, ethertype IPv4 (0x0800), length 342: (tos 0xc0, ttl 64, id 28145, offset 0, flags [none], proto UDP (17), length 328)
    172.19.31.1.67 > 172.19.31.96.68: BOOTP/DHCP, Reply, length 300, xid 0x628fd85b, Flags [none]
          Your-IP 172.19.31.96
          Server-IP 172.19.31.1
          Client-Ethernet-Address 00:00:6c:09:4c:d0
          Vendor-rfc1048 Extensions
            Magic Cookie 0x63825363
            DHCP-Message (53), length 1: Offer
            Server-ID (54), length 4: 172.19.31.1
            Lease-Time (51), length 4: 3600
            RN (58), length 4: 1800
            RB (59), length 4: 3150
            BR (28), length 4: 172.19.31.255
            Domain-Name-Server (6), length 4: 1.1.1.1
            Default-Gateway (3), length 4: 172.19.31.1
            Subnet-Mask (1), length 4: 255.255.255.0
08:16:09.276488 00:00:6c:09:4c:d0 > ff:ff:ff:ff:ff:ff, ethertype IPv4 (0x0800), length 342: (tos 0x0, ttl 64, id 0, offset 0, flags [none], proto UDP (17), length 328)
    0.0.0.0.68 > 255.255.255.255.67: BOOTP/DHCP, Request from 00:00:6c:09:4c:d0, length 300, xid 0x628fd85b, Flags [none]
          Client-Ethernet-Address 00:00:6c:09:4c:d0
          Vendor-rfc1048 Extensions
            Magic Cookie 0x63825363
            Requested-IP (50), length 4: 172.19.31.96
            DHCP-Message (53), length 1: Request
            Server-ID (54), length 4: 172.19.31.1
            Parameter-Request (55), length 4:
              Subnet-Mask (1), Default-Gateway (3), Domain-Name (15), Domain-Name-Server (6)
            MSZ (57), length 2: 1500
08:16:09.291305 4a:1f:7b:fb:b6:8e > 00:00:6c:09:4c:d0, ethertype IPv4 (0x0800), length 342: (tos 0xc0, ttl 64, id 28146, offset 0, flags [none], proto UDP (17), length 328)
    172.19.31.1.67 > 172.19.31.96.68: BOOTP/DHCP, Reply, length 300, xid 0x628fd85b, Flags [none]
          Your-IP 172.19.31.96
          Server-IP 172.19.31.1
          Client-Ethernet-Address 00:00:6c:09:4c:d0
          Vendor-rfc1048 Extensions
            Magic Cookie 0x63825363
            DHCP-Message (53), length 1: ACK
            Server-ID (54), length 4: 172.19.31.1
            Lease-Time (51), length 4: 3600
            RN (58), length 4: 1800
            RB (59), length 4: 3150
            BR (28), length 4: 172.19.31.255
            Domain-Name-Server (6), length 4: 1.1.1.1
            Default-Gateway (3), length 4: 172.19.31.1
            Subnet-Mask (1), length 4: 255.255.255.0
```

The rebinding procedure succeeds, so the load balancer retains its original IP address. This implies that the DHCP client can contact the DHCP server; the network condition is in good shape, so we can exclude the networking problem. With that said, there is no apparent defect from the user's point of view, but we should keep an eye on it.

**To Reproduce**

1. Prepare a v1.4-head Harvester cluster
2. Ensure there is a working DHCP server on the management network
3. Create a virtual machine
4. Create a load balancer (DHCP IPAM) for SSH port 22 associated with the created virtual machine
5. The created load balancer should get an IP address on the management network from the DHCP server
6. Wait for the lease renewal to happen (It depends on the DHCP server configuration. If the lease expiration time is set to an hour, the renewal is expected to start 30 minutes after the lease was created.)
7. The renewal should have failed (observed from the kube-vip leader pod's log)
8. The subsequent (if the lease expiration time is an hour, this should happen at the 52.5-minute mark) rebinding should succeed

**Expected behavior**

The DHCP lease renewal procedure should succeed if the network connectivity between the DHCP server and the client is good.

**Environment**

 - Harvester ISO version: v1.4-head (`85538537`)
 - Underlying Infrastructure (e.g. Baremetal with Dell PowerEdge R630): KVM
 - kube-vip version: v0.8.1

**Additional context**

An interesting finding: There is a [long-pending PR](https://github.com/insomniacslk/dhcp/pull/458) on which kube-vip depends. Due to the inability to merge the PR, kube-vip depended on [the forked one instead](https://github.com/kube-vip/kube-vip/pull/351/files#diff-33ef32bf6c23acb95f5902d7097b7a1d5128ca061167ec0716715b0b9eeaa5f6R5). However, this temporary Go module replacement was removed from [this PR](https://github.com/kube-vip/kube-vip/pull/587#issuecomment-1660167311). This means it's no longer the case, starting with the kube-vip v0.6.1 release. It is suspected to be relevant to the potential issue we raised here.

**Comment by innobead** [comment]
Regards this a upstream issue. Moved to a later version, and let's revisit this later. 

cc @harvester/tpm

**Comment by Mithrandir2k18** [comment]
We currently seem affected by this bug. It seems fixed upstream. Am I correct to assume that this bug being assigned to Milestone 1.6.0 means that the ETA of the fix is also 2025.7.25? or will harvester v1.4 get the fix sooner?

Using IPPools to assign reserved IPs to loadbalancers would be a workaround, correct?

**Comment by starbops** [comment]
Hi @Mithrandir2k18, thank you for reaching out.

> Am I correct to assume that this bug being assigned to Milestone 1.6.0 means that the ETA of the fix is also 2025.7.25? or will harvester v1.4 get the fix sooner?

If the bug has been fixed upstream, Harvester still needs to bump the kube-vip version. The schedule is currently TBD.

> Using IPPools to assign reserved IPs to loadbalancers would be a workaround, correct?

Yes, as this is a DHCP-specific issue, using IPPools should be fine.
